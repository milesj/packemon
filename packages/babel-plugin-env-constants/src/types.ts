/* eslint-disable no-underscore-dangle */

declare global {
	const __DEV__: boolean;
	const __PROD__: boolean;
	const __TEST__: boolean;
}

export type {};
