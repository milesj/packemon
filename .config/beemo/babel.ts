const plugins: object[] = [];

if (!process.env.NO_INTEROP) {
	// Needed for Jest
	plugins.push(['babel-plugin-cjs-esm-interop', { format: 'cjs' }]);
}

export default {
	plugins,
};
