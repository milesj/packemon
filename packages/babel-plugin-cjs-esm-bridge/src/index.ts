import paths from 'path';
import { NodePath, PluginObj, types as t } from '@babel/core';

function isEsmFile(file: string) {
	return /\.(ts|tsx|mjs)$/.test(file);
}

export default function cjsEsmBridge(): PluginObj {
	return {
		visitor: {
			CallExpression: {
				enter(path: NodePath<t.CallExpression>, state) {
					const isEsm = isEsmFile(state.filename);

					// require() not allowed in esm files
					// https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
					if (isEsm && path.get('callee').isIdentifier({ name: 'require' })) {
						throw new Error(
							`Found a \`require()\` call in "${paths.basename(
								state.filename,
							)}", this is not allowed in ".mjs" files. Use dynamic \`import()\` instead.`,
						);
					}
				},
			},
		},
	};
}
