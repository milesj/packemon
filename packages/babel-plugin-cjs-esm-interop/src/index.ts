/* eslint-disable complexity */

import paths from 'path';
import { NodePath, PluginObj, PluginPass, types as t } from '@babel/core';
// @ts-expect-error Not typed
import { addDefault } from '@babel/helper-module-imports';

export interface CjsEsmInteropOptions {
	// The output format
	format?: 'cjs' | 'mjs';
}

// function isEsmFile(file: string) {
// 	return /\.(ts|tsx|mjs)$/.test(file);
// }

function isPathDirname(path: NodePath): boolean {
	// dirname(foo)
	if (path.isCallExpression() && path.get('callee').isIdentifier({ name: 'dirname' })) {
		return true;
	}

	// path.dirname(foo)
	return Boolean(
		path.isCallExpression() &&
			path.get('callee').isMemberExpression() &&
			(path.get('callee.object') as NodePath).isIdentifier() &&
			(path.get('callee.property') as NodePath).isIdentifier({ name: 'dirname' }),
	);
}

function isProcessEnv(path: NodePath): boolean {
	// process.env
	return Boolean(
		path.isMemberExpression() &&
			(path.get('object') as NodePath).isMemberExpression() &&
			(path.get('object.object') as NodePath).isIdentifier({ name: 'process' }) &&
			(path.get('object.property') as NodePath).isIdentifier({ name: 'env' }),
	);
}

function getFormat(state: PluginPass): 'cjs' | 'mjs' {
	return (state.opts as CjsEsmInteropOptions)?.format ?? 'mjs';
}

function wrapWithFileProto(path: NodePath, id: string) {
	if (
		path.parentPath?.isNewExpression() &&
		path.parentPath?.get('callee').isIdentifier({ name: 'URL' })
	) {
		return t.binaryExpression('+', t.stringLiteral('file://'), t.identifier(id));
	}

	return t.identifier(id);
}

export default function cjsEsmInterop(): PluginObj {
	return {
		visitor: {
			CallExpression(path: NodePath<t.CallExpression>, state) {
				// // `require()` not allowed in esm files
				// // https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
				// if (getFormat(state) === 'mjs' && path.get('callee').isIdentifier({ name: 'require' })) {
				// 	throw new Error(
				// 		`Found a \`require()\` call in non-module file "${paths.basename(
				// 			state.filename,
				// 		)}". Use dynamic \`import()\` instead.`,
				// 	);
				// }

				// `require.resolve()` not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_resolve
				if (
					getFormat(state) === 'mjs' &&
					path.get('callee').isMemberExpression() &&
					(path.get('callee.object') as NodePath).isIdentifier({ name: 'require' }) &&
					(path.get('callee.property') as NodePath).isIdentifier({ name: 'resolve' })
				) {
					throw new Error(
						`Found a \`require.resolve()\` call in non-module file "${paths.basename(
							state.filename,
						)}". Use the \`resolve\` npm package instead.`,
					);
				}
			},

			Identifier(path: NodePath<t.Identifier>, state) {
				// __filename -> import.meta.url
				// https://nodejs.org/api/esm.html#esm_no_filename_or_dirname
				if (getFormat(state) === 'mjs' && path.isIdentifier({ name: '__filename' })) {
					path.replaceWith(
						t.memberExpression(
							t.metaProperty(t.identifier('import'), t.identifier('meta')),
							t.identifier('url'),
						),
					);
				}

				// __dirname -> path.dirname(import.meta.url)
				// https://nodejs.org/api/esm.html#esm_no_filename_or_dirname
				if (getFormat(state) === 'mjs' && path.isIdentifier({ name: '__dirname' })) {
					// eslint-disable-next-line @typescript-eslint/no-unsafe-call
					this.pathImport ??= addDefault(path, 'path', { nameHint: '_path' });

					const call = t.callExpression(
						t.memberExpression(
							t.identifier((this.pathImport as { name: string }).name),
							t.identifier('dirname'),
						),
						[
							t.memberExpression(
								t.metaProperty(t.identifier('import'), t.identifier('meta')),
								t.identifier('url'),
							),
						],
					);

					path.replaceWith(call);
				}

				// `NODE_PATH` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_node_path
				if (
					getFormat(state) === 'mjs' &&
					path.isIdentifier({ name: 'NODE_PATH' }) &&
					isProcessEnv(path.parentPath)
				) {
					throw new Error(
						'Environment variable `process.env.NODE_PATH` is not available in modules.',
					);
				}
			},

			MemberExpression(path: NodePath<t.MemberExpression>, state) {
				// const isEsm = isEsmFile(state.filename);
				// const file = paths.basename(state.filename);

				// // `exports.<name>` not allowed in esm files
				// // https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
				// if (
				// 	isEsm &&
				// 	getFormat(state) === 'mjs' &&
				// 	path.get('object').isIdentifier({ name: 'exports' }) &&
				// 	path.get('property').isIdentifier()
				// ) {
				// 	console.log(state, this);
				// 	const { name } = path.node.property as t.Identifier;

				// 	throw new Error(
				// 		`Found an \`exports.${name} =\` expression in non-module file "${file}". Use \`export const ${name} =\` instead.`,
				// 	);
				// }

				// // `module.exports` not allowed in esm files
				// // https://nodejs.org/api/esm.html#esm_no_require_exports_or_module_exports
				// if (
				// 	isEsm &&
				// 	getFormat(state) === 'mjs' &&
				// 	path.get('object').isIdentifier({ name: 'module' }) &&
				// 	path.get('property').isIdentifier({ name: 'exports' })
				// ) {
				// 	console.log(state, this);
				// 	throw new Error(
				// 		`Found a \`module.exports =\` expression in non-module file "${file}". Use \`export default\` instead.`,
				// 	);
				// }

				// import.meta.url -> __filename
				// path.dirname(import.meta.url) -> __dirname
				// https://nodejs.org/api/esm.html#esm_no_filename_or_dirname
				if (
					getFormat(state) === 'cjs' &&
					path.get('object').isMetaProperty() &&
					(path.get('object.meta') as NodePath).isIdentifier({ name: 'import' }) &&
					(path.get('object.property') as NodePath).isIdentifier({ name: 'meta' }) &&
					path.get('property').isIdentifier({ name: 'url' })
				) {
					if (isPathDirname(path.parentPath)) {
						path.parentPath.replaceWith(wrapWithFileProto(path.parentPath, '__dirname'));
					} else {
						path.replaceWith(wrapWithFileProto(path, '__filename'));
					}
				}

				// `require.extensions` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_extensions
				if (
					getFormat(state) === 'mjs' &&
					path.get('object').isIdentifier({ name: 'require' }) &&
					path.get('property').isIdentifier({ name: 'extensions' })
				) {
					throw new Error('API `require.extensions` is not available in modules.');
				}

				// `require.cache` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_cache
				if (
					getFormat(state) === 'mjs' &&
					path.get('object').isIdentifier({ name: 'require' }) &&
					path.get('property').isIdentifier({ name: 'cache' })
				) {
					throw new Error('API `require.cache` is not available in modules.');
				}

				// `require.resolve` is not allowed in esm files
				// https://nodejs.org/api/esm.html#esm_no_require_resolve
				if (
					getFormat(state) === 'mjs' &&
					path.get('object').isIdentifier({ name: 'require' }) &&
					path.get('property').isIdentifier({ name: 'resolve' })
				) {
					throw new Error(
						`Found a \`require.resolve()\` call in non-module file "${paths.basename(
							state.filename,
						)}". Use the \`resolve\` npm package instead.`,
					);
				}
			},
		},
	};
}
