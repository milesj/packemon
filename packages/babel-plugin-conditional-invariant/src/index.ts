import { NodePath, types as t } from '@babel/core';

function isWrappedWithConditional(path: NodePath<t.ExpressionStatement>): boolean {
	let currentPath: NodePath<t.Node> | null = path;
	let wrapped = false;

	while (currentPath) {
		// Is there a cleaner way of doing this???
		if (
			currentPath.isIfStatement() &&
			currentPath.get('test').isBinaryExpression({ operator: '!==' }) &&
			(currentPath.get('test.left') as NodePath<t.MemberExpression>).isMemberExpression() &&
			(currentPath.get('test.left.object') as NodePath<t.MemberExpression>).isMemberExpression() &&
			(currentPath.get('test.left.object.object') as NodePath<t.Identifier>).isIdentifier({
				name: 'process',
			}) &&
			(currentPath.get('test.left.object.property') as NodePath<t.Identifier>).isIdentifier({
				name: 'env',
			}) &&
			(currentPath.get('test.left.property') as NodePath<t.Identifier>).isIdentifier({
				name: 'NODE_ENV',
			}) &&
			(currentPath.get('test.right') as NodePath<t.StringLiteral>).isStringLiteral({
				value: 'production',
			})
		) {
			wrapped = true;
			break;
		}

		currentPath = currentPath.parentPath;
	}

	return wrapped;
}

export default function conditionalInvariantPlugin() {
	return {
		visitor: {
			ExpressionStatement: {
				enter(path: NodePath<t.ExpressionStatement>) {
					if (
						!path.get('expression').isCallExpression() ||
						!(path.get('expression.callee') as NodePath<t.Identifier>).isIdentifier({
							name: 'invariant',
						})
					) {
						return;
					}

					// Only transform invariants within the following scenarios:
					// 	- Block statements (if, while, functions, etc)
					// 	- Switch case statements
					//	- Module scope
					if (
						!path.parentPath.isBlockStatement() &&
						!path.parentPath.isSwitchCase() &&
						!path.parentPath.isProgram()
					) {
						return;
					}

					// Dont replace with new syntax if invariant is already wrapped
					// in a NODE_ENV conditional!
					if (isWrappedWithConditional(path)) {
						return;
					}

					path.replaceWith(
						t.ifStatement(
							t.binaryExpression(
								'!==',
								t.memberExpression(
									t.memberExpression(t.identifier('process'), t.identifier('env'), false),
									t.identifier('NODE_ENV'),
									false,
								),
								t.stringLiteral('production'),
							),
							t.blockStatement([path.node]),
						),
					);
				},
			},
		},
	};
}
