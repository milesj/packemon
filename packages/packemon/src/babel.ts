import fs from 'fs-extra';
import { TransformOptions as ConfigStructure } from '@babel/core';
import { Path, toArray } from '@boost/common';
import { FeatureFlags } from './types';
import {
	CodeArtifact,
	DEFAULT_SUPPORT,
	Format,
	getBabelInputConfig,
	getBabelOutputConfig,
	Package,
	PackemonPackage,
	Platform,
	Project,
	Support,
} from '.';

const format = (process.env.PACKEMON_FORMAT ?? 'lib') as Format;
const support = (process.env.PACKEMON_SUPPORT ?? DEFAULT_SUPPORT) as Support;
const project = new Project(process.cwd());

function getBabelConfig(artifact: CodeArtifact, featureFlags: FeatureFlags): ConfigStructure {
	const inputConfig = getBabelInputConfig(artifact, featureFlags);
	const outputConfig = getBabelOutputConfig(
		artifact.platform,
		artifact.support,
		artifact.builds[0].format,
		featureFlags,
	);

	return {
		// Input must come first
		plugins: [...inputConfig.plugins!, ...outputConfig.plugins!],
		// Env must come first
		presets: [...outputConfig.presets!, ...inputConfig.presets!],
	};
}

export interface ConfigOptions {
	format?: Format;
	platform?: Platform;
	support?: Support;
}

export function createConfig(folder: string, options: ConfigOptions = {}): ConfigStructure {
	const path = new Path(folder);
	const contents = fs.readJsonSync(path.append('package.json').path()) as PackemonPackage;

	// Create package and configs
	const pkg = new Package(project, path, contents);

	if (pkg.packageJson.packemon) {
		pkg.setConfigs(toArray(pkg.packageJson.packemon));
	}

	// Determine the lowest platform to support
	const platforms = pkg.configs.map((config) => config.platform);
	let lowestPlatform: Platform = 'node';

	// istanbul ignore next
	if (platforms.includes('browser')) {
		lowestPlatform = 'browser';
	} else if (platforms.includes('native')) {
		lowestPlatform = 'native';
	}

	// Generate artifact and builds
	const artifact = new CodeArtifact(pkg, [{ format: options.format ?? format }]);
	artifact.bundle = false;
	artifact.platform = options.platform ?? lowestPlatform;
	artifact.support = options.support ?? support;

	return getBabelConfig(artifact, pkg.getFeatureFlags());
}

export function createRootConfig(options?: ConfigOptions): ConfigStructure {
	const config = createConfig(process.cwd(), options);

	return {
		...config,
		babelrc: true,
		babelrcRoots: project.getWorkspaceGlobs({ relative: true }),
		// Support React Native libraries by default
		overrides: [
			{
				presets: ['@babel/preset-flow'],
				test: /node_modules\/((jest-)?react-native|@react-native(-community)?)/iu,
			},
		],
	};
}
