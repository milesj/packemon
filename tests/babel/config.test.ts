import { getBabelInputConfig, getBabelOutputConfig } from '../../src/babel/config';
import { Format, Platform, Support } from '../../src/types';

jest.mock('../../src/babel/resolve', () => (name) => name);

const SUPPORTS: Support[] = ['legacy', 'stable', 'current', 'experimental'];

function renderPresetEnv(platform: Platform, format: Format, support: Support) {
  it(`handles preset-env: ${platform} + ${format} + ${support}`, () => {
    expect(getBabelOutputConfig({ format, platform, support }, {}).presets[0]).toMatchSnapshot();
  });
}

describe('getBabelOutputConfig()', () => {
  SUPPORTS.forEach((support) => {
    renderPresetEnv('native', 'lib', support);

    (['lib', 'cjs', 'mjs'] as const).forEach((format) => {
      renderPresetEnv('node', format, support);
    });

    (['lib', 'esm', 'umd'] as const).forEach((format) => {
      renderPresetEnv('browser', format, support);
    });
  });

  it('errors for invalid platform', () => {
    expect(() =>
      // @ts-expect-error
      getBabelOutputConfig({ format: 'lib', platform: 'unknown', support: 'stable' }, {}),
    ).toThrow('Unknown platform "unknown".');
  });

  it('transforms async/await to promises when `browser` or `native`', () => {
    expect(
      getBabelOutputConfig({ format: 'lib', platform: 'browser', support: 'stable' }, {}),
    ).toMatchSnapshot();

    expect(
      getBabelOutputConfig({ format: 'lib', platform: 'native', support: 'experimental' }, {}),
    ).toMatchSnapshot();
  });

  it('uses built-in destructuring and object spread when `current` or `experimental`', () => {
    expect(
      getBabelOutputConfig({ format: 'lib', platform: 'node', support: 'current' }, {}),
    ).toMatchSnapshot();

    expect(
      getBabelOutputConfig({ format: 'lib', platform: 'node', support: 'experimental' }, {}),
    ).toMatchSnapshot();
  });

  it('sets `parserOpts.strictMode` based on `strict` feature flag', () => {
    expect(
      getBabelOutputConfig(
        { format: 'lib', platform: 'node', support: 'stable' },
        { strict: true },
      ),
    ).toEqual(
      expect.objectContaining({
        parserOpts: {
          sourceType: 'unambiguous',
          strictMode: true,
        },
      }),
    );
  });

  it('sets `babelrcRoots` based on `workspaces` feature flag', () => {
    expect(
      getBabelOutputConfig(
        { format: 'lib', platform: 'node', support: 'stable' },
        { workspaces: ['packages/*'] },
      ),
    ).toEqual(
      expect.objectContaining({
        babelrcRoots: ['packages/*'],
      }),
    );
  });
});
