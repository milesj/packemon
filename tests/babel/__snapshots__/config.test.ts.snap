// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBabelOutputConfig() handles preset-env: browser + esm + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "defaults",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + esm + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "last 2 chrome versions",
        "last 2 firefox versions",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + esm + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 10",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + esm + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 11",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "defaults",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "last 2 chrome versions",
        "last 2 firefox versions",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 10",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + lib + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 11",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "defaults",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": Array [
        "last 2 chrome versions",
        "last 2 firefox versions",
      ],
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 10",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: browser + umd + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "umd",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "IE 11",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 12",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 14",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 8",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: native + lib + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "browsers": "iOS 10",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "12.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "15.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "8.10.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + cjs + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "10.3.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "12.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "15.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "8.10.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + lib + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": "cjs",
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "10.3.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + current 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "12.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + experimental 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "15.0.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + legacy 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "8.10.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() handles preset-env: node + mjs + stable 1`] = `
Array [
  "@babel/preset-env",
  Object {
    "bugfixes": true,
    "exclude": Array [
      "@babel/plugin-transform-regenerator",
      "@babel/plugin-transform-async-to-generator",
    ],
    "loose": false,
    "modules": false,
    "shippedProposals": true,
    "spec": true,
    "targets": Object {
      "node": "10.3.0",
    },
    "useBuiltIns": false,
  },
]
`;

exports[`getBabelOutputConfig() transforms async/await to promises when \`browser\` or \`native\` 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": false,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "babel-plugin-transform-async-to-promises",
      Object {
        "inlineHelpers": true,
        "target": "es5",
      },
    ],
    Object {
      "visitor": Object {
        "Identifier": Object {
          "enter": [Function],
        },
      },
    },
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "browsers": "IE 11",
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;

exports[`getBabelOutputConfig() transforms async/await to promises when \`browser\` or \`native\` 2`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": false,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "@babel/plugin-transform-destructuring",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "babel-plugin-transform-async-to-promises",
      Object {
        "inlineHelpers": true,
        "target": "es6",
      },
    ],
    Object {
      "visitor": Object {
        "Identifier": Object {
          "enter": [Function],
        },
      },
    },
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "browsers": "iOS 14",
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;

exports[`getBabelOutputConfig() uses built-in destructuring and object spread when \`current\` or \`experimental\` 1`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": false,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "@babel/plugin-transform-destructuring",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    Object {
      "visitor": Object {
        "Identifier": Object {
          "enter": [Function],
        },
      },
    },
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "exclude": Array [
          "@babel/plugin-transform-regenerator",
          "@babel/plugin-transform-async-to-generator",
        ],
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "node": "12.0.0",
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;

exports[`getBabelOutputConfig() uses built-in destructuring and object spread when \`current\` or \`experimental\` 2`] = `
Object {
  "babelrc": true,
  "babelrcRoots": undefined,
  "caller": Object {
    "name": "packemon",
  },
  "comments": false,
  "configFile": false,
  "parserOpts": Object {
    "sourceType": "unambiguous",
    "strictMode": undefined,
  },
  "plugins": Array [
    Array [
      "@babel/plugin-transform-destructuring",
      Object {
        "useBuiltIns": true,
      },
    ],
    Array [
      "@babel/plugin-proposal-object-rest-spread",
      Object {
        "useBuiltIns": true,
      },
    ],
    Object {
      "visitor": Object {
        "Identifier": Object {
          "enter": [Function],
        },
      },
    },
  ],
  "presets": Array [
    Array [
      "@babel/preset-env",
      Object {
        "bugfixes": true,
        "exclude": Array [
          "@babel/plugin-transform-regenerator",
          "@babel/plugin-transform-async-to-generator",
        ],
        "loose": false,
        "modules": "cjs",
        "shippedProposals": true,
        "spec": true,
        "targets": Object {
          "node": "15.0.0",
        },
        "useBuiltIns": false,
      },
    ],
  ],
}
`;
